### ISSUE TITLE
Send data with VM remote_write from vmagent to vminsert

### ISSUE THREAD

**Issue by LoricAndre** [VictoriaMetrics/VictoriaMetrics#6085]
### Is your question request related to a specific component?

vmagent, vminsert

### Describe the question in detail

First, thank you for this amazing product, everything has been crystal clear so far.

However, I am currently trying to send data from a vmagent instance to a VM cluster, and [the docs](https://docs.victoriametrics.com/vmagent/#victoriametrics-remote-write-protocol) seem to indicate that this is possible, but I can't find the `remoteWrite.url` to use in the list [here](https://docs.victoriametrics.com/#how-to-import-time-series-data).

What URL am I supposed to use ?

### Troubleshooting docs

- [X] General - https://docs.victoriametrics.com/Troubleshooting.html
- [X] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [X] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by jiekun** [comment]
Please check out the docs of VMCluster: https://docs.victoriametrics.com/cluster-victoriametrics/#url-format

And one example of vmagent remote write config could be:
```yaml
  remoteWrite:
    - url: "http://my.k8s.service.url:8480/insert/0/prometheus/api/v1/write"
```

**Comment by LoricAndre** [comment]
Thank you for the answer, I am aware of those differences and tried the urls with that format. However, sending data using the URL you indicated does not seem to work.
What I am doing to test if the endpoint supports VM remote write and not prometheus' one is adding the `remoteWrite.forceVMProto` flag. When sending to your endpoint, vminsert returns a status 400 with a `snappy: corrupt input`, which seems coherent with the endpoint expecting prometheus remote_write data.

**Comment by jiekun** [comment]
Hi. Have you try it without the `forceVMProto` flag?

I think the endpoint's name `/prometheus/xxx` is a little bit confusing, but they could accept vm protocol and I tested myself.

Here's the related code block of `vminsert` and `vmagent`:
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/a63e8c46dc1809cbf637e959876b64aec7308cca/app/vminsert/main.go#L167
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/a63e8c46dc1809cbf637e959876b64aec7308cca/app/vmagent/remotewrite/client.go#L159

If no flag is set (both `remoteWrite.forceVMProto` and `remoteWrite.forcePromProto`), `vmagent` will try to negotiate with `vminsert` during **init** and decide what protocol to use.

So I suggest:
- Test again without `remoteWrite.forceVMProto` flag to see if auto-negotiate works well;
- For vm protocol, it would be nice if you could provide full config you're using, also the version of related vm components. I will be happy to test and see if other compression settings affected it.

**Comment by LoricAndre** [comment]
Thank you very much for taking the time to help me.

I tried removing the forceVMProto:
```
victoria_metrics  | {"ts":"2024-04-12T08:48:43.156Z","level":"info","caller":"VictoriaMetrics/lib/logger/flag.go:12","msg":"build version: vmagent-20240116-231749-tags-v1.93.10-0-gd277977e7"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.156Z","level":"info","caller":"VictoriaMetrics/lib/logger/flag.go:13","msg":"command-line flags"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.156Z","level":"info","caller":"VictoriaMetrics/lib/logger/flag.go:20","msg":"  -loggerFormat=\"json\""}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.156Z","level":"info","caller":"VictoriaMetrics/lib/logger/flag.go:20","msg":"  -promscrape.config=\"./prometheus.yaml\""}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.156Z","level":"info","caller":"VictoriaMetrics/lib/logger/flag.go:20","msg":"  -remoteWrite.label=\"instance=<redacted>\""}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.156Z","level":"info","caller":"VictoriaMetrics/lib/logger/flag.go:20","msg":"  -remoteWrite.tmpDataPath=\"/vmagent-remotewrite-data\""}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.156Z","level":"info","caller":"VictoriaMetrics/lib/logger/flag.go:20","msg":"  -remoteWrite.url=\"secret\""}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.156Z","level":"info","caller":"VictoriaMetrics/app/vmagent/main.go:120","msg":"starting vmagent at \":8429\"..."}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.157Z","level":"info","caller":"VictoriaMetrics/lib/memory/memory.go:42","msg":"limiting caches to 9823653888 bytes, leaving 6549102592 bytes to the OS accor
ding to -memory.allowedPercent=60"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.158Z","level":"info","caller":"VictoriaMetrics/lib/persistentqueue/fastqueue.go:60","msg":"opened fast persistent queue at \"/vmagent-remotewrite-data/persi
stent-queue/1_5A52A1E0778C9703\" with maxInmemoryBlocks=3200, it contains 0 pending bytes"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.305Z","level":"info","caller":"VictoriaMetrics/app/vmagent/remotewrite/client.go:163","msg":"the remote storage at \"1:secret-url\" doesn't support Victoria
Metrics remote write protocol. Switching to Prometheus remote write protocol. See https://docs.victoriametrics.com/vmagent.html#victoriametrics-remote-write-protocol"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.305Z","level":"info","caller":"VictoriaMetrics/app/vmagent/remotewrite/client.go:200","msg":"initialized client for -remoteWrite.url=\"1:secret-url\""}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.306Z","level":"info","caller":"VictoriaMetrics/app/vmagent/main.go:146","msg":"started vmagent in 0.149 seconds"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.306Z","level":"info","caller":"VictoriaMetrics/lib/promscrape/scraper.go:113","msg":"reading Prometheus configs from \"./prometheus.yaml\""}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.306Z","level":"info","caller":"VictoriaMetrics/lib/httpserver/httpserver.go:96","msg":"starting http server at http://127.0.0.1:8429/"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.306Z","level":"info","caller":"VictoriaMetrics/lib/httpserver/httpserver.go:97","msg":"pprof handlers are exposed at http://127.0.0.1:8429/debug/pprof/"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.308Z","level":"info","caller":"VictoriaMetrics/lib/promscrape/config.go:122","msg":"starting service discovery routines..."}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.308Z","level":"info","caller":"VictoriaMetrics/lib/promscrape/config.go:128","msg":"started service discovery routines in 0.000 seconds"}
victoria_metrics  | {"ts":"2024-04-12T08:48:43.310Z","level":"info","caller":"VictoriaMetrics/lib/promscrape/scraper.go:430","msg":"static_configs: added targets: 4, removed targets: 0; total targets: 4"
}
```

My vmagent is version 1.93.10, started with:
```yaml
  vmagent:
    image: victoriametrics/vmagent:v1.93.10
    container_name: victoria_metrics
    volumes:
      - "./data/vmagent:/vmagent-remotewrite-data"
    configs:
      - source: prometheus_config
        target: "./prometheus.yaml"
    entrypoint:
      - "/vmagent-prod"
      - "-remoteWrite.url=${VM_SERVER_WRITE_ENDPOINT}"
      - "-remoteWrite.label=instance=<redacted>}"
      - "-promscrape.config=./prometheus.yaml"
      - "-remoteWrite.tmpDataPath=/vmagent-remotewrite-data"
      - "-loggerFormat=json"

```
And my env vars:
```sh
VM_TENANT_ID="0"
VM_SERVER_WRITE_ENDPOINT="https://metrics.$BACKEND_DOMAIN/insert/$VM_TENANT_ID/prometheus/api/v1/write"
```

VMInsert is in a k3s cluster, reachable through the network and then through nginx-ingress-controller.
The values for victoria-metrics-cluster are:
```yaml
rbac:
  pspEnabled: false
vmselect:
  statefulSet:
    enabled: true
  serviceMonitor:
    enabled: true
vminsert:
  ingress:
    enabled: true
    className: nginx
    annotations:
      kubernetes.io/tls-acme: "true"
  serviceMonitor:
    enabled: true
---
vmselect:
  replicaCount: 1
vmstorage:
  retention: 6month
  size: 8Gi
  replicaCount: 1
vminsert:
  replicaCount: 1
  ingress:
    hosts:
      - name: &host metrics.<redacted>
        path: /insert
        port: http
    tls:
      - secretName: prometheus-pushgateway-tls-secret
        hosts:
          - *host
```

Is there anything obviously misconfigured in here ? Or could it be nginx ?

**Comment by jiekun** [comment]
The `vmagent` log indicated that remote write target does not support vm protocol:
```
victoria_metrics  | {"ts":"2024-04-12T08:48:43.305Z","level":"info","caller":"VictoriaMetrics/app/vmagent/remotewrite/client.go:163","msg":"the remote storage at \"1:secret-url\" doesn't support Victoria Metrics remote write protocol. Switching to Prometheus remote write protocol. See https://docs.victoriametrics.com/vmagent.html#victoriametrics-remote-write-protocol"}
```

Could you have a quick glance at the version of `vminsert` (and probably other components of vmcluster)?

**Comment by LoricAndre** [comment]
Sure, the chart version is **0.9.16**, with all 3 components in version **1.73.0**

**Comment by LoricAndre** [comment]
All right, upgrading the cluster to the latest chart version (0.11.14) seems to have fixed it. Really sorry about all that, and thank you for the help !

**Comment by jiekun** [comment]
That's correct. 

According to this comment: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1225#issuecomment-1437824333
the new protocol is introduced in last year, probably. but anyway it's not included in v1.73.0.
Also see: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/v1.73.0/lib/protoparser/promremotewrite/streamparser.go#L36
